### ISSUE TITLE
Segmentation fault on G++ when trying to assign json string literal to custom json type.

### ISSUE THREAD

**Issue by zaycakitayca** [nlohmann/json#972]
**Bug Report**

This is the minimal code demonstrating the problem.

```cpp
#include <iostream>
#include "json.hpp"
#include "fifo_map.hpp"

template<class K, class V, class dummy_compare, class A>
using my_workaround_fifo_map = nlohmann::fifo_map<K, V, nlohmann::fifo_map_compare<K>, A>;
using my_json = nlohmann::basic_json<my_workaround_fifo_map>;

int main()
{
	my_json foo = R"([1, 2, 3])"_json;
	
	std::cout << foo << std::endl;
}
```

```
$ g++ --version
g++ (GCC) 7.3.0
$ uname -a
Linux daddy 4.14.15-1-ARCH #1 SMP PREEMPT Tue Jan 23 21:49:25 UTC 2018 x86_64 GNU/Linux
```

**Comment by nlohmann** [comment]
Out of curiosity: does it work without fifo_map?

**Comment by zaycakitayca** [comment]
Do you mean `nlohmann::json` instead of `my_json` in the code above? Yes, it works just fine.

**Comment by nlohmann** [comment]
I'll check tonight.

**Comment by gregmarr** [comment]
Do you have a stack trace?  Does it happen with just the assignment line, or does it require the `cout`?

**Comment by zaycakitayca** [comment]
`cout` here is just to suppress unused variable warning, the program actually fails on assignment. I can run this programin debugger tomorrow and attach its output here if it's required.

**Comment by gregmarr** [comment]
Just thought that if you had it handy, we might be able to see something before Niels gets to look at it tonight.

**Comment by nlohmann** [comment]
When I call `my_json foo = my_json::parse("[1, 2, 3]");` instead of `my_json foo = R"([1, 2, 3])"_json;`, the error disappears. The problems seems to be the definition of `_json` which does not return `basic_json`, but `json`.

[stacktrace.txt](https://github.com/nlohmann/json/files/1725291/stacktrace.txt)

The error can be reduced to an assignment from `json` to the user-defined version:

```cpp
#include "json.hpp"
#include "fifo_map.hpp"

template<class K, class V, class dummy_compare, class A>
using my_workaround_fifo_map = nlohmann::fifo_map<K, V, nlohmann::fifo_map_compare<K>, A>;
using my_json = nlohmann::basic_json<my_workaround_fifo_map>;

int main()
{
    nlohmann::json x = 1;
    my_json y = x;
}
```

**Comment by nlohmann** [comment]
I think the problem is that the copy constructor/assignment operator/whatever thinks assigning a `json` object to `nlohmann::basic_json<my_workaround_fifo_map>` is a good idea. I think this should be avoided in the first place.

**Comment by nlohmann** [comment]
In #986 there is a fix for a related issue. Maybe it also helps here.

**Comment by nlohmann** [comment]
With PR #986, the original code example compiles and runs without issues. Output is `[1,2,3]`.

**Comment by nlohmann** [comment]
Fixed with #986.
